File: python_sdk/src/reality_capture/specifications/change_detection.py
from pydantic import BaseModel, Field
from typing import Optional
from enum import Enum


class ChangeDetectionInputs(BaseModel):
    model_3d_a: str = Field(alias="model3DA", description="Reality data id of ContextScene, point cloud, Gaussian splats, or mesh")
    model_3d_b: str = Field(alias="model3DB", description="Reality data id of ContextScene, point cloud, Gaussian splats, or mesh")
    extent: Optional[str] = Field(None, alias="extent", pattern=r"^bkt:.+",
                                  description="Path in the bucket of the clipping polygon to apply")
    preset: Optional[str] = Field(None, alias="preset", pattern=r"^bkt:.+",
                                  description="Path in the bucket of a preset file to use")


class ChangeDetectionOutputs(BaseModel):
    segmentation_3d_a: Optional[str] = Field(None, description="Reality data id of ContextScene, pointing to the segmented 3D model A", alias="segmentation3DA")
    segmented_model_3d_b: Optional[str] = Field(None, description="Reality data id of the 3D segmented model in the same format of model 3d B", alias="segmentedModel3DB")
    segmentation_3d_b: Optional[str] = Field(None, description="Reality data id of ContextScene, pointing to the segmented 3D model B", alias="segmentation3DB")
    segmented_model_3d_a: Optional[str] = Field(None, description="Reality data id of the 3D segmented model in the same format of model 3d A", alias="segmentedModel3DA")
    locations3d_a: Optional[str] = Field(None, alias="locations3DA",
                                              description="Reality data id of ContextScene with locations of changes in A")
    locations3d_a_as_shp: Optional[str] = Field(None, alias="locations3DAAsSHP",
                                              description="Reality data id of locations of changes in A"
                                                          "as SHP format")
    locations3d_a_as_geojson: Optional[str] = Field(None, alias="locations3DAAsGeoJSON",
                                                  description="Reality data id of locations of changes in A"
                                                              "as GeoJSON file")
    locations3d_b: Optional[str] = Field(None, alias="locations3DB",
                                         description="Reality data id of ContextScene with locations of changes in B")
    locations3d_b_as_shp: Optional[str] = Field(None, alias="locations3DBAsSHP",
                                                description="Reality data id of locations of changes in B as SHP format")
    locations3d_b_as_geojson: Optional[str] = Field(None, alias="locations3DBAsGeoJSON",
                                                    description="Reality data id of locations of changes in B as GeoJSON file")
    


class ChangeDetectionOutputsCreate(Enum):
    SEGMENTATION3D_A = "segmentation3DA"
    SEGMENTED_MODEL3D_A = "segmentedModel3DA"
    SEGMENTATION3D_B = "segmentation3DB"
    SEGMENTED_MODEL3D_B = "segmentedModel3DB"
    LOCATIONS3D_A = "locations3DA"
    LOCATIONS3D_A_AS_SHP = "locations3DAAsSHP"
    LOCATIONS3D_A_AS_GEOJSON = "locations3DAAsGeoJSON"
    LOCATIONS3D_B = "locations3DB"
    LOCATIONS3D_B_AS_SHP = "locations3DBAsSHP"
    LOCATIONS3D_B_AS_GEOJSON = "locations3DBAsGeoJSON"
    

class ChangeDetectionOptions(BaseModel):
    min_points_per_change: Optional[int] = Field(None, alias="minPointsPerChange",
                                                 description="Minimum number of points in a region "
                                                             "to be considered as a change")
    sampling_resolution: Optional[float] = Field(None, alias="samplingResolution",
                                                 description="Target point cloud resolution when starting from meshes")
    grow_threshold: Optional[float] = Field(None, alias="growThreshold",
                                            description="Low threshold to detect spatial changes "
                                                         "(hysteresis detection)")
    filter_threshold: Optional[float] = Field(None, alias="filterThreshold",
                                              description="High threshold to detect spatial changes "
                                                          "(hysteresis detection)")


class ChangeDetectionSpecificationsCreate(BaseModel):
    inputs: ChangeDetectionInputs = Field(description="Inputs")
    outputs: list[ChangeDetectionOutputsCreate] = Field(description="Outputs")
    options: Optional[ChangeDetectionOptions] = Field(None, description="Options")


class ChangeDetectionSpecifications(BaseModel):
    inputs: ChangeDetectionInputs = Field(description="Inputs")
    outputs: ChangeDetectionOutputs = Field(description="Outputs")
    options: Optional[ChangeDetectionOptions] = Field(None, description="Options")

File: typescript/packages/reality-capture/src/specifications/change_detection.ts
import { z } from "zod";

export enum ChangeDetectionOutputsCreate {
  SEGMENTATION3D_A = "segmentation3DA",
  SEGMENTED_MODEL3D_A = "segmentedModel3DA",
  SEGMENTATION3D_B = "segmentation3DB",
  SEGMENTED_MODEL3D_B = "segmentedModel3DB",
  LOCATIONS3D_A = "locations3DA",
  LOCATIONS3D_A_AS_SHP = "locations3DAAsSHP",
  LOCATIONS3D_A_AS_GEOJSON = "locations3DAAsGeoJSON",
  LOCATIONS3D_B = "locations3DB",
  LOCATIONS3D_B_AS_SHP = "locations3DBAsSHP",
  LOCATIONS3D_B_AS_GEOJSON = "locations3DBAsGeoJSON"
}

export const ChangeDetectionInputsSchema = z.object({
  model3DA: z.string().describe("Reality data id of ContextScene, point cloud, GS or mesh"),
  model3DB: z.string().describe("Reality data id of ContextScene, point cloud, GS or mesh"),
  extent: z.string().describe("Path in the bucket of the clipping polygon to apply").regex(/^bkt:.+/, {
    message: "Path in the bucket to the extent file must start with 'bkt:'",
  }).optional(),
  // Add preset in a different release
  /*preset: z.string().describe("Path in the bucket of a preset file to use").regex(/^bkt:.+/, {
    message: "Path in the bucket to the preset file must start with 'bkt:'",
  }).optional(),*/
});
export type ChangeDetectionInputs = z.infer<typeof ChangeDetectionInputsSchema>;

export const ChangeDetectionOutputsSchema = z.object({
  segmentation3DA: z.string().optional().describe("ContextScene CD on Model3D A"),
  segmentedModel3DA: z.string().optional().describe("Model3D A with CD segmentation"),
  segmentation3DB: z.string().optional().describe("ContextScene CD on Model3D B"),
  segmentedModel3DB: z.string().optional().describe("Model3D B with CD segmentation"),
  locations3DA: z.string().optional().describe("Reality data id of locations of changes A"),
  locations3DAAsSHP: z.string().optional().describe("Reality data id of locations of changes A as SHP format"),
  locations3DAAsGeoJSON: z.string().optional().describe("Reality data id of locations of changes A as GeoJSON file"),
  locations3DB: z.string().optional().describe("Reality data id of locations of changes B"),
  locations3DBAsSHP: z.string().optional().describe("Reality data id of locations of changes B as SHP format"),
  locations3DBAsGeoJSON: z.string().optional().describe("Reality data id of locations of changes B as GeoJSON file"),
});
export type ChangeDetectionOutputs = z.infer<typeof ChangeDetectionOutputsSchema>;

export const ChangeDetectionOptionsSchema = z.object({
  minPointsPerChange: z.number().int().optional().describe("Minimum number of points in a region to be considered as a change"),
  samplingResolution: z.number().optional().describe("Target point cloud resolution when starting from meshes"),
  growThreshold: z.number().optional().describe("Low threshold to detect spatial changes (hysteresis detection)"),
  filterThreshold: z.number().optional().describe("High threshold to detect spatial changes (hysteresis detection)"),
});
export type ChangeDetectionOptions = z.infer<typeof ChangeDetectionOptionsSchema>;

export const ChangeDetectionSpecificationsCreateSchema = z.object({
  inputs: ChangeDetectionInputsSchema.describe("Inputs"),
  outputs: z.array(z.nativeEnum(ChangeDetectionOutputsCreate)).describe("Outputs"),
  options: ChangeDetectionOptionsSchema.optional().describe("Options"),
});
export type ChangeDetectionSpecificationsCreate = z.infer<typeof ChangeDetectionSpecificationsCreateSchema>;

export const ChangeDetectionSpecificationsSchema = z.object({
  inputs: ChangeDetectionInputsSchema.describe("Inputs"),
  outputs: ChangeDetectionOutputsSchema.describe("Outputs"),
  options: ChangeDetectionOptionsSchema.optional().describe("Options"),
});
export type ChangeDetectionSpecifications = z.infer<typeof ChangeDetectionSpecificationsSchema>;